keps/sig-cluster-lifecycle/kubeadm/4214-separate-super-user-kubeconfig/README.md

```diff
@@ -0,0 +1,724 @@
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+To preserve the existing behavior where `admin.conf` has full cluster access,
+a new ClusterRoleBinding will be created called `kubeadm:cluster-admins`.
+It will bind the ClusterRole `cluster-admin` to the `kubeadm:cluster-admins` Group.
+The credential stored in `admin.conf` will have the following subject:
+`O = kubeadm:cluster-admins, CN = kubernetes-admin`. In case of a compromised
+credential, the ClusterRoleBinding `kubeadm:cluster-admins` can be removed or updated.
+The Group `kubeadm:cluster-admins` is recommended for internal kubeadm use only.
+
+The new file `super-admin.conf` will contain the following subject:
+`O = system:masters, CN = kubernetes-super-admin`. It will act as the "break-glass",
+super-user credential that can bypass RBAC. In case this credential is compromised
+the cluster certificate authority must be rotated.
+
+Both the `admin.conf` and `super-admin.conf` files will be renewable by `kubeadm upgrade`
+and `kubeadm certs renew`. If the `super-admin.conf` is missing it will not cause an error.
+That is in case the super-admin has manually moved the file to a safe location -
+i.e. not keeping it on the primary control plane node, where `kubeadm init` was called.
+
+The documentation in the
+[Generating kubeconfig files for additional users](https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-certs/#kubeconfig-additional-users)
+section will be updated.
+For each new User or a Group of users the recommendation will be to create a new
+ClusterRoleBinding. Use the flags `--client-name` and `--org` of `kubeadm kubeconfig user`
+to control what User or Group this credential belongs to. Revoke access of a single User
+or a whole Group in case of credential compromise. Using the existing `system:masters` or
+`kubeadm:cluster-admins` Groups will not be recommended.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story: Resolving compromised admin credentials
+
+As a super-user administrator I want to be able to use RBAC to remove access of an
+administrator credential that has been compromised. By removing the `kubeadm:cluster-admins`
+ClusterRoleBinding all administrators credentials (`admin.conf`) on control plane nodes
+that are signed for the `kubeadm:cluster-admins` Group will stop working.
+I can then proceed to rotate the certificate authority of this cluster and sign new
```

Comment:
I wonder if we can use a configurable Group name rather than the static `kubeadm:cluster-admins` Group. If this is possible, the administrator only needs to revoke the original Group and use a new Group name to re-generate an `admin.conf` instead of rotating the certificate authority and all the certs of the cluster. It's more convenient and low-cost. 

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: kubeadm:cluster-admins
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: cluster-admin
subjects:
- apiGroup: rbac.authorization.k8s.io
  kind: Group
  name: <configurable-here>
```

Reply:
technically the administrators can just delete the kubeadm:cluster-admins group and bind new admin.conf files to a new group. that would also work. the above comment in the doc that talks about rotating the CA, is with precaution in mind - i.e. they don't need to do it but it's a recommendation, because there is no certificate revocation.

i think we can make the group name configurable, but even if we do the steps will remain the same. 1. create new group, 2. update bindings. 3. create new admin.conf files, 4. optionally rotate CA.

maybe this rotation of CA can be marked as optional in the doc?

Reply:
+1 mark rotation of CA as optional

Reply:
updated. added it as an alternative route. users can just choose to not rotate CA and create a new group, remove the compromised CRB, as noted above.